src/func.cc

```diff
@@ -656,6 +657,59 @@ void ShellFuncNoRerun(const std::vector<Value*>& args,
   ShellStatusVar::SetValue(returnCode);
 }
 
+void VarVisibilityFunc(const std::vector<Value*>& args,
+                       Evaluator* ev,
+                       std::string* s) {
+  std::string arg = args[0]->Eval(ev);
+  std::vector<std::string> prefixes;
+
+  std::stringstream ss(args[1]->Eval(ev));
+  std::string prefix;
+  while (ss >> prefix) {
+    if (HasPrefix(prefix, "/")) {
+      ERROR_LOC(ev->loc(), "Visibility prefix should not start with /");
+    }
+    if (HasPrefix(prefix, "../")) {
```

Comment:
This allows "./../", which has the same effect. The best solution is probably to just not print normalizedPrefix in line 681, rather than trying to detect all of the places where the passed argument would escape the workspace.

Reply:
I think this is fine, `./../` would fail the normalization check.

Reply:
What is the purpose of having this check here? If it was removed completely the normalization check would still fail.

This check is either:
 - redundant because of the normalization check, or
 - insufficient in its effort to avoid leaking the parent directory name in the normalization check. If leaking the parent directory is considered bad, then just drop that from the error message below.

In any case, this check provides no added value.

Reply:
`NormalizePath` can still produce paths starting with `../`. For example, `../foo` is a normalized path. But a `../` anywhere else in the path would be removed by the normalization process, e.g. `foo/../../bar` normalizes to `../bar`, so we'd catch it.

I don't think this check is either redundant or insufficient.

It's not really about leaking the parent directory name, it's that you'd never want to make a variable visible to a location outside of the android source tree, so we should deny that nonsense argument.

Reply:
I was misunderstanding `NormalizePath`, and expecting "../foo" to turn into an absolute path.

Agreed.